server/src/main/java/io/crate/planner/optimizer/symbol/Optimizer.java

```diff
@@ -132,14 +135,22 @@ public Symbol tryApplyRules(Symbol node) {
     }
 
     private class Visitor extends FunctionCopyVisitor<Void> {
+        private Stack<io.crate.expression.symbol.Function> visitedFunctions = new Stack<>();
```

Comment:
could be replace by a counter since the only usage right now is to check if the current node is the root or not.

Reply:
Would at least change it to `ArrayDeque` to avoid the synchronization overhead that `Stack` has.